src/Microsoft.Framework.DependencyInjection.Abstractions/ServiceDescriptor.cs

```diff
@@ -72,6 +72,24 @@ private ServiceDescriptor(Type serviceType, ServiceLifetime lifetime)
         /// <inheritdoc />
         public Func<IServiceProvider, object> ImplementationFactory { get; }
 
+        public Type GetImplementationType()
+        {
+            if (ImplementationType != null)
+            {
+                return ImplementationType;
+            }
+            else if (ImplementationInstance != null)
+            {
+                return ImplementationInstance.GetType();
+            }
+            else if (ImplementationFactory != null)
+            {
+                return ImplementationFactory.GetType().GenericTypeArguments[1];
```

Comment:
`ImplementationFactory` is always a `Func<IServiceProvider, object>`, so this should be safe.

I wonder if we should throw an exception if the returned type would be `object`.